pkg/dumper/postgres/dumper.go

```diff
@@ -23,119 +22,72 @@ type pgDumper struct {
-	return &pgDumper{
-		conn:   conn,
-		reader: rdr,
-	}
+	return generic.NewSqlDumper(
+		rdr,
+		&pgDumper{
+			conn:   conn,
+			reader: rdr,
+		},
+	)
 }
 
-func (p *pgDumper) Dump(done chan<- struct{}, configTables config.Tables) error {
-	if err := p.dumpStructure(); err != nil {
+func (p *pgDumper) DumpStructure(sql string) error {
+	if _, err := p.conn.Exec(sql); err != nil {
 		return err
 	}
 
-	return p.dumpTables(done, configTables)
-}
-
-func (p *pgDumper) Close() error {
-	return p.conn.Close()
+	return nil
 }
 
-func (p *pgDumper) dumpStructure() error {
-	log.Debug("Dumping structure...")
-	structureSQL, err := p.reader.GetStructure()
+func (p *pgDumper) DumpTable(tableName string, rowChan <-chan database.Row) error {
+	txn, err := p.conn.Begin()
 	if err != nil {
-		return errors.Wrap(err, "failed to read structure")
+		return errors.Wrap(err, "failed to open transaction")
 	}
 
-	_, err = p.conn.Exec(structureSQL)
+	insertedRows, err := p.insertIntoTable(txn, tableName, rowChan)
 	if err != nil {
-		return errors.Wrap(err, "failed to dump structure")
+		return errors.Wrap(err, "failed to insert rows")
 	}
+	log.WithField("table", tableName).WithField("inserted", insertedRows).Debug("inserted rows")
 
-	log.Debug("Structure dumped")
-	return nil
-}
-
-func (p *pgDumper) dumpTables(done chan<- struct{}, configTables config.Tables) error {
-	// Get the tables
-	tables, err := p.reader.GetTables()
-	if err != nil {
-		return err
+	if err := txn.Commit(); err != nil {
+		return errors.Wrap(err, "failed to commit transaction")
 	}
 
-	if err := p.disableTriggers(tables); err != nil {
-		return err
-	}
+	return nil
+}
 
-	var wg sync.WaitGroup
-	wg.Add(len(tables))
+// disableTriggers Disable triggers on all tables
+func (p *pgDumper) PreDumpTables(tables []string) error {
+	// We can't use `SET session_replication_role = replica` because multiple connections and stuff
 	for _, tbl := range tables {
-		logger := log.WithField("table", tbl)
-		logger.Info("Dumping table data...")
-		var opts reader.ReadTableOpt
-
-		table, err := configTables.FindByName(tbl)
-		if err != nil {
-			logger.WithError(err).Debug("no configuration found for table")
-		}
-
-		if table != nil {
-			opts = reader.ReadTableOpt{
-				Limit:         table.Filter.Limit,
-				Relationships: p.relationshipConfigToOptions(table.Relationships),
-			}
-		}
-
-		// Create read/write chanel
-		rowChan := make(chan database.Row)
-
-		go func(tableName string, rowChan <-chan database.Row) {
-			if err := p.dumpTable(tableName, rowChan); err != nil {
-				logger.WithError(err).Error("Failed to dump table")
-			}
-
-			wg.Done()
-			logger.Info("Done dumping table data")
-		}(tbl, rowChan)
-
-		if err := p.reader.ReadTable(tbl, rowChan, opts); err != nil {
-			logger.WithError(err).Error("error while reading table")
+		query := fmt.Sprintf("ALTER TABLE %s DISABLE TRIGGER ALL", strconv.Quote(tbl))
+		if _, err := p.conn.Exec(query); err != nil {
```

Comment:
Have you considered to run this concurrently rather than serial? I think we can do this soon

Reply:
I'm going to leave this for now. We can optimize this when it becomes needed.

Reply:
Agreed 👍